### DISCUSSION TITLE
abi3 module cross compilation without python interpreter

### DISCUSSION THREAD

**Discussion by emanuele-f** [PyO3/pyo3#2310] [Questions] [answered]
I'm trying to cross compile an abi3 module without the burden to cross-compile the python interpreter, as explained in https://pyo3.rs/main/building_and_distribution.html. I'm setting the following environment variables:

```
PYO3_NO_PYTHON=1
PYO3_CROSS_PYTHON_VERSION="3.8.7"
PYO3_CROSS_LIB_DIR=/usr/lib/python3.8
```

The `/usr/lib/python3.8` contains the `_sysconfigdata__linux_x86_64-linux-gnu.py` file for the building machine. However, the build fails with the following error: `attempted to link to Python shared library but config does not contain lib_name`. What am I missing?

Here is my environment:

```
Python: 3.8.7
      platform: Linux-5.17.1-arch1-1-x86_64-with-glibc2.35
      pip: 20.2.3
      setuptools: 62.1.0
      setuptools_rust: 1.2.0
```

**Comment by adamreichold** [comment]
Not sure whether this is causing the failure you experience, but `PYO3_CROSS_PYTHON_VERSION` should consist of only major and minor version AFAIK, e.g. `export PYO3_CROSS_PYTHON_VERSION=3.8`.

**Reply by emanuele-f** [reply]
Tried with 3.8 but still failing

**Reply by adamreichold** [reply]
Since this was apparently accepted, which version of PyO3 are you using by the way?

**Reply by emanuele-f** [reply]
pyo3 v0.15.1

**Reply by adamreichold** [reply]
Just for completeness because this area of the code base has seen significant improvements in the last release, could you give 0.16.4 a try?

**Comment by adamreichold** [comment, marked as answer]
For debugging purposes, could you set the environment variable `PYO3_PRINT_CONFIG` and share the result here? (This will fail the build which is intentional. But the output captured by Cargo should contain the inferred configuration.)

**Reply by emanuele-f** [reply]
Cryptography 36.0.2

```
    -- PYO3_PRINT_CONFIG=1 is set, printing configuration and halting compile --
    implementation=CPython
    version=3.6
    shared=true
    abi3=true
    build_flags=WITH_THREAD
    suppress_build_script_link_lines=false
```

Additional environment:

```
RUSTFLAGS="-C linker=/root/target/toolchains/x86_64/bin/x86_64-linux-android-gcc"
CARGO_BUILD_TARGET="x86_64-linux-android"
```

**Reply by adamreichold** [reply]
I fear that what you want to achieve might not yet be possible in PyO3 0.15.1, c.f. the guide matching that version: https://pyo3.rs/v0.15.1/building_and_distribution.html

**Reply by emanuele-f** [reply]
Ok, I think this is the problem. pyo3 0.16.4 cross compiles correctly now but the cryptography 36.0.2 rust module compilation fails, probably because it depends on abi3-py36 which was dropped in 0.16.0. Thank you for your help uncovering this.

Related discussion: https://github.com/pyca/cryptography/issues/6924#issuecomment-1056978447
